src/v/s3/client.cc

```diff
@@ -777,4 +772,31 @@ void client_pool::release(ss::shared_ptr<client> leased) {
     _cvar.signal();
 }
 
+void client_pool::load_credentials(cloud_roles::credentials&& credentials) {
+    if (unlikely(!_apply_credentials)) {
+        _apply_credentials = ss::make_lw_shared(
+          cloud_roles::make_credentials_applier(std::move(credentials)));
+        init();
+        // We signal the waiter only after the client pool is initialized, so
+        // that any upload operations waiting are ready to proceed.
+        _credentials_var.signal();
+    } else {
+        _apply_credentials->reset_creds(std::move(credentials));
+    }
+}
+
+ss::future<> client_pool::wait_for_credentials() {
+    while (!_apply_credentials && !_gate.is_closed()
```

Comment:
You don't need a while loop here if you are using wait method that takes a predicate. You can check all conditions in the predicate.

Reply:
fixed